File: core/src/promitech/colonization/ai/WanderMissionHandler.java
package promitech.colonization.ai;

import java.util.ArrayList;
import java.util.List;

import net.sf.freecol.common.model.Game;
import net.sf.freecol.common.model.Settlement;
import net.sf.freecol.common.model.Tile;
import net.sf.freecol.common.model.ai.missions.PlayerMissionsContainer;
import net.sf.freecol.common.model.ai.missions.WanderMission;
import net.sf.freecol.common.model.player.Player;
import promitech.colonization.Direction;
import promitech.colonization.Randomizer;
import promitech.colonization.SpiralIterator;
import promitech.colonization.move.MoveContext;
import promitech.colonization.move.MoveService;
import promitech.map.Boolean2dArray;

public class WanderMissionHandler implements MissionHandler<WanderMission> {

	private final Boolean2dArray settlementWanderRange;
	private Player wanderRangeForPlayer;
	
	private final Game game;
	private final MoveService moveService;
	
	private final List<Direction> allowedDirections = new ArrayList<Direction>(Direction.allDirections);
	private final MoveContext moveContext = new MoveContext();
	
	public WanderMissionHandler(Game game, MoveService moveService) {
		this.game = game;
		this.moveService = moveService;
		this.settlementWanderRange = new Boolean2dArray(game.map.width, game.map.height);
	}

	private void prepareSettlementWanderRange(Player player) {
		if (wanderRangeForPlayer != null && wanderRangeForPlayer.equalsId(player)) {
			return;
		}
		wanderRangeForPlayer = player;
		settlementWanderRange.set(false);
		SpiralIterator spiral = new SpiralIterator(game.map.width, game.map.height);
		for (Settlement s : player.settlements.entities()) {
			spiral.reset(s.tile.x, s.tile.y, true, s.settlementType.getWanderingRadius());
			while (spiral.hasNext()) {
				settlementWanderRange.set(spiral.getX(), spiral.getY(), true);
				spiral.next();
			}
		}
	}
	
	@Override
	public void handle(PlayerMissionsContainer playerMissionsContainer, WanderMission mission) {
		prepareSettlementWanderRange(mission.unit.getOwner());
		
		Tile sourceTile = mission.unit.getTile();
		
		boolean canMove = false;
		do {
			canMove = false;
			
			Direction moveDirection = generateRandomDirection(mission, sourceTile);
			if (moveDirection == null) {
				mission.previewDirection = null;
				return;
			}
			
			Tile destTile = game.map.getTile(sourceTile, moveDirection);
			moveContext.init(sourceTile, destTile, mission.unit, moveDirection);
			
			if (moveContext.canHandleMove()) {
	            moveService.aiConfirmedMoveProcessor(moveContext);
				
				mission.previewDirection = moveDirection;					
				
				if (mission.unit.hasMovesPoints()) {
					canMove = true;
					sourceTile = destTile;
				}
			} else {
				canMove = false;
			}
		} while (canMove);
	}
	
	private Direction generateRandomDirection(final WanderMission mission, final Tile sourceTile) {
		Direction comeDirection = null;
		if (mission.previewDirection != null) {
			comeDirection = mission.previewDirection.getReverseDirection();
		}
		allowedDirections.clear();
		for (Direction direction : Direction.allDirections) {
			Tile neighbourTile = game.map.getTile(sourceTile, direction);
			if (neighbourTile == null) {
				continue;
			}
			if (comeDirection != null && comeDirection == direction) {
				continue;
			}
			if (!settlementWanderRange.get(neighbourTile.x, neighbourTile.y)) {
				continue;
			}
			moveContext.init(sourceTile, neighbourTile, mission.unit, direction);
			
			if (moveContext.canHandleMove()) {
				allowedDirections.add(direction);
			}
		}
		
		if (allowedDirections.isEmpty()) {
			return null;
		}
		return Randomizer.instance().randomMember(allowedDirections);
	}
	
}

File: core/src/net/sf/freecol/common/model/ai/missions/indian/WanderMissionHandler.kt
package net.sf.freecol.common.model.ai.missions.indian

import net.sf.freecol.common.model.Game
import net.sf.freecol.common.model.MoveType
import net.sf.freecol.common.model.Tile
import net.sf.freecol.common.model.Unit
import net.sf.freecol.common.model.ai.missions.PlayerMissionsContainer
import net.sf.freecol.common.model.map.path.Path
import net.sf.freecol.common.model.map.path.PathFinder
import net.sf.freecol.common.model.player.Player
import promitech.colonization.Direction
import promitech.colonization.Randomizer
import promitech.colonization.ai.CommonMissionHandler
import promitech.colonization.ai.MissionHandler
import promitech.colonization.ai.MissionHandlerLogger.logger
import promitech.colonization.orders.combat.CombatService
import promitech.colonization.orders.move.MoveContext
import promitech.colonization.orders.move.MoveService

class WanderMissionHandler(
    val game: Game,
    val moveService: MoveService,
    val pathFinder: PathFinder,
    val combatService: CombatService
) : MissionHandler<WanderMission> {

    private val settlementWanderRange = SettlementWanderRange(game.map)

    private val allowedDirections: MutableList<Direction> = ArrayList(Direction.allDirections)
    private val moveContext = MoveContext()

    override fun handle(playerMissionsContainer: PlayerMissionsContainer, mission: WanderMission) {
        val player = playerMissionsContainer.player
        val unit: Unit? = player.units.getByIdOrNull(mission.unitId)

        if (unit == null || !CommonMissionHandler.isUnitExists(player, unit)) {
            mission.setDone()
            return
        }
        settlementWanderRange.prepare(unit.owner)

        while (true) {
            val aggressionTile = AttackDecision.determineTileToAggression(game.map, unit)
            if (aggressionTile != null) {
                moveToAggressionTile(mission, aggressionTile, player, unit)
                break
            }
            if (!moveToRandomDirection(mission, unit)) {
                break
            }
        }
    }

    private fun moveToAggressionTile(mission: WanderMission, aggressionTile: Tile, player: Player, unit: Unit) {
        if (logger.isDebug) {
            logger.debug("player[%s].wanderMissionHandler[%s].tileToAggression[%s]",
                player.id,
                mission.id,
                aggressionTile.toStringCords()
            )
        }
        val pathToEnemy: Path = pathFinder.findToTile(game.map, unit.tile, aggressionTile, unit, PathFinder.includeUnexploredTiles)
        if (pathToEnemy.isReachedDestination) {
            val moveContext = MoveContext(unit, pathToEnemy)
            moveContext.initNextPathStep()
            while (moveContext.canAiHandleMove()) {
                if (moveContext.isMoveType(MoveType.ATTACK_UNIT) || moveContext.isMoveType(MoveType.ATTACK_SETTLEMENT)) {
                    combatService.doCombat(moveContext)
                    if (unit.isDisposed) {
                        mission.setDone()
                    }
                    break;
                }
                moveService.aiConfirmedMoveProcessor(moveContext)
                moveContext.initNextPathStep()
            }
        }
    }

    private fun moveToRandomDirection(mission: WanderMission, unit: Unit): Boolean {
        val sourceTile = unit.tile
        val moveDirection = generateRandomDirection(mission, sourceTile, unit)
        if (moveDirection == null) {
            mission.previewDirection = null
            return false
        }
        val destTile = game.map.getTile(sourceTile, moveDirection)
        moveContext.init(sourceTile, destTile, unit, moveDirection)
        if (moveContext.canAiHandleMove()) {
            moveService.aiConfirmedMoveProcessor(moveContext)
            mission.previewDirection = moveDirection
            return unit.hasMovesPoints()
        }
        return false
    }

    private fun generateRandomDirection(mission: WanderMission, sourceTile: Tile, unit: Unit): Direction? {
        val comeDirection: Direction? = mission.previewReverseDirection()

        allowedDirections.clear()

        for (direction in Direction.allDirections) {
            val neighbourTile = game.map.getTile(sourceTile, direction) ?: continue
            if (comeDirection != null && comeDirection == direction) {
                continue
            }
            if (!settlementWanderRange.inRange(neighbourTile)) {
                continue
            }

            moveContext.init(sourceTile, neighbourTile, unit, direction)
            if (moveContext.canAiHandleMove() && moveContext.isMoveType) {
                allowedDirections.add(direction)
            }
        }
        return if (allowedDirections.isEmpty()) {
            null
        } else Randomizer.instance().randomMember(allowedDirections)
    }

}